### PR TITLE
Move rustfmt downloads from bootstrap.py to rustbuild

### LINKED ISSUE #95136: Using `cargo run` with bootstrap uses the host rustfmt, not nightly rustfmt
https://github.com/rust-lang/rust/blob/c7ce69faf2a7ea16c15d922985ca27ba70da30ee/src/bootstrap/config.rs#L1095-L1106

The proper fix is to check if we're going through the `cargo run` entrypoint instead of x.py, and if so, download rustfmt from the nightly toolchain listed in `src/stage0.json`: https://github.com/rust-lang/rust/blob/c7ce69faf2a7ea16c15d922985ca27ba70da30ee/src/stage0.json#L8-L11

Ideally we would eventually be able to move this into rustup, but that requires not only for rustup to add and release this feature, but also for all contributors to start using it. It shouldn't be too much extra work for bootstrap to do it itself in the meantime.

cc https://github.com/rust-lang/rust/issues/94829, https://github.com/rust-lang/rust/pull/94806#issuecomment-1073257368, https://github.com/rust-lang/rust/pull/94830#issuecomment-1066807203

### PR DESCRIPTION
**Pull request by jyn514**
- Allow verifying CI downloads using src/stage0.json
- Change download functions not to hard-code `ci-artifacts.rust-lang.org`
- Change `format::format` to take a `Builder` so it has access to `download_component`. I think we may want to reconsider the distinction between Build and Builder at some point; I don't think it's particularly useful.
- Move rustfmt downloads out of bootstrap.py

Fixes https://github.com/rust-lang/rust/issues/95136. Helps with https://github.com/rust-lang/rust/issues/94829. This is based on https://github.com/rust-lang/rust/pull/96687 for simplicity.

### CODE REVIEW

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -63,31 +63,30 @@ def support_xz():
...
-                        print("using already-download file", path)
-                    return
-                else:
-                    if verbose:
-                        print("ignoring already-download file",
-                            path, "due to failed verification")
-                    os.unlink(path)
+        if url not in checksums:
+            raise RuntimeError(("src/stage0.json doesn't contain a checksum for {}. "
+                                "Pre-built artifacts might not available for this "
```
Comment (est31) [inline src/bootstrap/bootstrap.py:73]: You are not changing the specific code here, but there is a word missing here :).

Comment (jyn514) [reply src/bootstrap/bootstrap.py]: even with your comment it took me about 3 minutes to figure out which word was missing :laughing: thanks!

**Comment by Mark-Simulacrum** [comment]
r=me with conflicts resolved

**Pull request merged**